File: Scala/algorithmic_toolbox/src/main/scala-2.12/introduction/HugeFibonacciModulo.scala
package introduction

import scala.annotation.tailrec

object HugeFibonacciModulo {
  private def convertToLongList(line: String): List[Long] = line.split(" ").map(_.toLong).toList

  private def calcPisanoPeriod(modulus: Int): Long = {
    @tailrec
    def loop(a: Int, b: Int, p: Long): Long = {
      if (a == 0 && b == 1) p
      else loop(b, (a + b) % modulus, p + 1)
    }
    loop(a = 1, b = 1, p = 1L)
  }

  private def calcFibonacciModulo(n: Long, modulus: Int): Int =
    (0L until n).foldLeft((0, 1)){ case ((a, b), _) => (b, (a + b) % modulus) }._1

  def calcHugeFibonacciModulo(n: Long, modulus: Int): Int = {
    val p: Long = calcPisanoPeriod(modulus)
    calcFibonacciModulo(n % p, modulus)
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val List(n, modulus): List[Long] = convertToLongList(reader.next())
    val result: Int = calcHugeFibonacciModulo(n, modulus.toInt)
    println(result)
  }
}

File: OCaml/algorithmic_toolbox/introduction/huge_fibonacci_modulo.ml
let calc_pisano_period modulus =
    let rec inner a b p =
        if a = 0 && b = 1 then p
        else inner b ((a + b) mod modulus) (p + 1)
    in inner 1 1 1


let calc_fibonacci_modulo n modulus =
    let rec inner k a b =
        if k < 1 then a
        else inner (k - 1) b ((a + b) mod modulus)
    in inner n 0 1


let calc_huge_fibonacci_modulo n modulus =
    let p = calc_pisano_period modulus
    in calc_fibonacci_modulo (n mod p) modulus


let () =
    let (n, modulus) = Scanf.scanf "%d %d" (fun x y -> (x, y)) in
    print_int (calc_huge_fibonacci_modulo n modulus);
    print_newline()
